File: mock-openai/mock-server.ts
#!/usr/bin/env -S deno run --allow-net
/// <reference lib="deno.ns" />
/// <reference lib="dom" />

const port = 8086;

const customResponseText = `
__Advertisement :)__

- __[pica](https://nodeca.github.io/pica/demo/)__ - high quality and fast image
  resize in browser.

You will like those projects!

---

# h1 Heading 8-)
## h2 Heading
### h3 Heading
#### h4 Heading
##### h5 Heading
###### h6 Heading

## Horizontal Rules

___

---

***

## Typographic replacements

Enable typographer option to see result.

(c) (C) (r) (R) (tm) (TM) (p) (P) +- 

test.. test... test..... test?..... test!....

!!!!!! ???? ,,  -- ---

"Smartypants, double quotes" and 'single quotes'
`;

function buildResponseTemplate() {
  return {
    choices: [
      {
        message: { content: customResponseText, role: "assistant" },
        finish_reason: "stop",
        index: 0,
        logprobs: null,
      },
    ],
    id: "chatcmpl-mock",
    model: "gpt-3.5-mock",
    object: "chat.completion",
    usage: { completion_tokens: 10, prompt_tokens: 9, total_tokens: 19 },
    created: Math.floor(Date.now() / 1000),
  };
}

async function handler(req: Request): Promise<Response> {
  const url = new URL(req.url);

  if (url.pathname === "/health" && req.method === "GET") {
    return new Response(JSON.stringify({ status: "ok" }), {
      status: 200,
      headers: { "Content-Type": "application/json" },
    });
  }

  if (url.pathname === "/v1/chat/completions" && req.method === "POST") {
    // parse body as unknown and narrow to a plain object map; avoid `any`
    const rawBody: unknown = await req.json().catch(() => ({}));
    const body: Record<string, unknown> = (
        typeof rawBody === "object" && rawBody !== null
      )
      ? (rawBody as Record<string, unknown>)
      : {};
    const responseTemplate = buildResponseTemplate();

    // If client asked for streaming, return SSE
    if ("stream" in body && Boolean(body["stream"])) {
      const content = responseTemplate.choices?.[0]?.message?.content || "";
      const words = content.split(/\s+/).filter(Boolean);
      const batchSize = 8;
      const latencyMs = 3; // very small delay for CI speed

      const stream = new ReadableStream({
        async start(controller) {
          for (let i = 0; i < words.length; i += batchSize) {
            const chunk = words.slice(i, i + batchSize).join(" ") +
              (i + batchSize < words.length ? " " : "");
            const payload = JSON.stringify({
              choices: [{ delta: { content: chunk } }],
            });
            controller.enqueue(
              new TextEncoder().encode(`data: ${payload}\n\n`),
            );
            // small delay
            await new Promise((r) => setTimeout(r, latencyMs));
          }
          controller.enqueue(new TextEncoder().encode("data: [DONE]\n\n"));
          controller.close();
        },
      });

      return new Response(stream, {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      });
    }

    // Non-streaming: return JSON immediately
    return new Response(JSON.stringify(responseTemplate), {
      status: 200,
      headers: { "Content-Type": "application/json" },
    });
  }

  return new Response("Not Found", { status: 404 });
}

export async function startMockServerProcess(): Promise<{ close: () => void }> {
  // Prefer spawning a subprocess when run permission is granted. Otherwise,
  // if the test runner granted network permission, start the server
  // in-process using Deno.serve and return a handle to stop it.
  try {
    const runPerm = await Deno.permissions.query({ name: "run" });
    if (runPerm.state === "granted") {
      const command = new Deno.Command(Deno.execPath(), {
        args: ["run", "--allow-net=127.0.0.1:8086", import.meta.url],
        stdout: "piped",
        stderr: "piped",
      });
      const proc = command.spawn();

      // Wait until the server is healthy or timeout
      const start = Date.now();
      while (Date.now() - start < 2000) {
        try {
          const res = await fetch("http://127.0.0.1:8086/health");
          try {
            await res.text();
          } catch {
            // ignore
          }
          if (res.ok) break;
        } catch (_e) {
          // server not up yet
        }
        await new Promise((r) => setTimeout(r, 10));
      }

      return {
        close: async () => {
          try {
            proc.kill();
          } catch (_e) {
            // ignore kill errors
          }

          // Close or cancel stdout/stderr if available to avoid leaks reported by
          // the Deno test runner. Different Deno versions expose slightly
          // different stream APIs, so try both cancel() and close() where
          // present.
          try {
            type StreamLike = {
              cancel?: () => Promise<void> | void;
              close?: () => void;
            };

            if (proc.stdout) {
              try {
                const s = proc.stdout as unknown as StreamLike;
                if (typeof s.cancel === "function") {
                  await s.cancel();
                }
              } catch {
                // ignore
              }
              try {
                const s = proc.stdout as unknown as StreamLike;
                if (typeof s.close === "function") {
                  s.close();
                }
              } catch {
                // ignore
              }
            }
            if (proc.stderr) {
              try {
                const s = proc.stderr as unknown as StreamLike;
                if (typeof s.cancel === "function") {
                  await s.cancel();
                }
              } catch {
                // ignore
              }
              try {
                const s = proc.stderr as unknown as StreamLike;
                if (typeof s.close === "function") {
                  s.close();
                }
              } catch {
                // ignore
              }
            }
          } catch {
            // ignore any close/cancel errors
          }

          try {
            // Wait for process to exit
            return await proc.status.catch(() => {});
          } catch (_e) {
            // ignore
            return Promise.resolve();
          }
        },
      };
    }
  } catch {
    // ignore permission query failures and fall through to net-based attempt
  }

  // If run isn't granted, try to start the server in-process when network
  // permission is available. This avoids requiring --allow-run for tests that
  // are allowed to bind to localhost.
  try {
    const netPerm = await Deno.permissions.query({
      name: "net",
      host: "127.0.0.1:8086",
    });
    if (netPerm.state === "granted") {
      const controller = new AbortController();
      // Start server in background; stop by aborting the controller
      Deno.serve(
        { hostname: "127.0.0.1", port, signal: controller.signal },
        handler,
      );

      // Wait for quick health check readiness
      const start = Date.now();
      while (Date.now() - start < 2000) {
        try {
          const res = await fetch("http://127.0.0.1:8086/health");
          if (res.ok) break;
        } catch {
          // not up yet
        }
        await new Promise((r) => setTimeout(r, 10));
      }

      return {
        close: () => {
          try {
            controller.abort();
          } catch {
            // ignore
          }
          return Promise.resolve();
        },
      };
    }
  } catch {
    // ignore permission query failures
  }

  throw new Error(
    "startMockServerProcess requires --allow-run or --allow-net=127.0.0.1:8086 to start the mock server",
  );
}

console.log(`Mock OpenAI API server module loaded at http://127.0.0.1:${port}`);
// If executed directly, start the server in-process.
if (import.meta.main) {
  console.log(`Mock OpenAI API server listening at http://127.0.0.1:${port}`);
  // Use the built-in Deno.serve to avoid deprecated std APIs.
  // Deno.serve will block the process and handle incoming requests.
  Deno.serve({ hostname: "127.0.0.1", port }, handler);
}

File: mock_server.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"strings"
)

// MockServer provides a simple mock OpenAI-compatible API for testing
type MockServer struct {
	server *http.Server
}

// ChatCompletionResponse represents the structure of a chat completion response
type ChatCompletionResponse struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Created int64  `json:"created"`
	Model   string `json:"model"`
	Choices []struct {
		Index   int `json:"index"`
		Message struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"message"`
		FinishReason string `json:"finish_reason"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage"`
}

// NewMockServer creates a new mock server
func NewMockServer(port string) *MockServer {
	mux := http.NewServeMux()
	mux.HandleFunc("/v1/chat/completions", handleChatCompletions)

	server := &http.Server{
		Addr:    ":" + port,
		Handler: mux,
	}

	return &MockServer{server: server}
}

// Start starts the mock server
func (m *MockServer) Start() error {
	log.Printf("Starting mock server on %s", m.server.Addr)
	return m.server.ListenAndServe()
}

// Stop stops the mock server
func (m *MockServer) Stop() error {
	return m.server.Close()
}

// handleChatCompletions handles chat completions requests
func handleChatCompletions(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// Parse the request
	var req ChatRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		return
	}

	// Extract the user message
	var userMessage string
	for _, msg := range req.Messages {
		if msg.Role == "user" {
			userMessage = msg.Content
			break
		}
	}

	// Generate a mock response based on the prompt
	responseContent := generateMockResponse(userMessage)

	// Create response
	response := ChatCompletionResponse{
		ID:      "chatcmpl-mock123",
		Object:  "chat.completion",
		Created: 1234567890,
		Model:   req.Model,
		Choices: []struct {
			Index   int `json:"index"`
			Message struct {
				Role    string `json:"role"`
				Content string `json:"content"`
			} `json:"message"`
			FinishReason string `json:"finish_reason"`
		}{
			{
				Index: 0,
				Message: struct {
					Role    string `json:"role"`
					Content string `json:"content"`
				}{
					Role:    "assistant",
					Content: responseContent,
				},
				FinishReason: "stop",
			},
		},
		Usage: struct {
			PromptTokens     int `json:"prompt_tokens"`
			CompletionTokens int `json:"completion_tokens"`
			TotalTokens      int `json:"total_tokens"`
		}{
			PromptTokens:     10,
			CompletionTokens: 20,
			TotalTokens:      30,
		},
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

// generateMockResponse generates a mock response based on the input
func generateMockResponse(prompt string) string {
	prompt = strings.ToLower(prompt)
	
	if strings.Contains(prompt, "hello") {
		return "Hello! How can I help you today?"
	}
	if strings.Contains(prompt, "test") {
		return "This is a test response from the mock server."
	}
	if strings.Contains(prompt, "markdown") {
		return "# Markdown Response\n\nThis is a **bold** response with *italic* text and `code`."
	}
	
	return "This is a mock response to: " + prompt
}